content/courses/advanced-ada/chapters/strings.rst

```diff
@@ -831,30 +832,370 @@ the :ada:`Decode` function to convert between the :ada:`UTF_16_Wide_String` and
 the :ada:`Wide_Wide_String` types.
 
 
-Universal text buffer
----------------------
+.. _Adv_Ada_Image_Attribute:
 
-.. admonition:: Relevant topics
+Image attribute
+---------------
 
-    - :arm22:`Universal Text Buffers <A-4-12>`
+Overview
+~~~~~~~~
 
-.. todo::
+In the :ref:`Introduction to Ada <Intro_Ada_Image_Attribute>` course, we've
+seen that the :ada:`'Image` attribute returns a string that contains a textual
+representation of an object. For example, we write :ada:`Integer'Image (V)` to
+get a string for the integer variable :ada:`V`:
 
-    Complete section!
+.. code:: ada run_button project=Courses.Advanced_Ada.Strings.Simple_Image
+
+    with Ada.Text_IO; use Ada.Text_IO;
 
+    procedure Show_Simple_Image is
+       V : Integer;
+    begin
+       V := 10;
+       Put_Line ("V: " & Integer'Image (V));
+    end Show_Simple_Image;
 
-Image attribute
----------------
+Naturally, we can use the :ada:`Image` attribute with other scalar types. For
+example:
 
-.. admonition:: Relevant topics
+.. code:: ada run_button project=Courses.Advanced_Ada.Strings.Simple_Image
 
-    - Image attribute mentioned in
-      :arm22:`Image Attributes <4-10>`
+    with Ada.Text_IO; use Ada.Text_IO;
 
-.. todo::
+    procedure Show_Simple_Image is
+       type Status is (Unknown, Off, On);
 
-    Complete section!
+       V : Float;
+       S : Status;
+    begin
+       V := 10.0;
+       S := Unknown;
+
+       Put_Line ("V: " & Float'Image (V));
+       Put_Line ("S: " & Status'Image (S));
+    end Show_Simple_Image;
+
+In this example, we retrieve a string representing the floating-point
+variable :ada:`V`. Also, we use :ada:`Status'Image (V)` to retrieve a string representing the textual version of the :ada:`Status`.
+
+.. admonition:: In the Ada Reference Manual
+
+    - :arm22:`Image Attributes <4-10>`
+
+
+:ada:`Type'Image` and :ada:`Obj'Image`
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We can also apply the :ada:`'Image` attribute to an object directly:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Strings.Simple_Image
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Show_Simple_Image is
+       V : Integer;
+    begin
+       V := 10;
+       Put_Line ("V: " & V'Image);
+
+       --  Equivalent to:
+       --  Put_Line ("V: " & Integer'Image (V));
+    end Show_Simple_Image;
+
+.. note::
+
+   This feature was introduced in Ada 2022.
```

Comment:
No, that was added by Ada 2012.

Reply:
Are you sure about it? When I look into the [corresponding section of the Ada 2022 AARM](http://www.ada-auth.org/standards/22aarm/html/AA-4-10.html), I see that `X'Image` was added to the standard. Also, when I look into the older Ada 2012 manual, I can find `S'Image` (for a subtype `S`), but not `X'Image`.

Reply:
In any case, I should be more specific and say:

> `Obj'Image` was introduced in Ada 2022.

instead of just saying

> This feature was introduced in Ada 2022.

Reply:
> The text was simplified (and thus the applicability of the attribute expanded) in Ada 2022, but X'Image was added in Ada 2012. Have a look at K.2 in particular for "X'Image".
> Or section 3.5 Scalar Types in Ada 2012 (because it applied only to objects of scalar types).

Yes, you're right: I (finally) found `X'Image` in the Ada 2012 standard. :-)

The reason for the confusion is that I was looking at the *original* 2012 standard:

http://www.ada-auth.org/standards/12rm/html/RM-3-5.html

This means, I was looking at the wrong place. `X'Image` was actually added to the 2016 corrections of the 2012 standard (Technical Corrigendum 1):

http://www.ada-auth.org/standards/rm12_w_tc1/html/RM-3-5.html

The Annotated Ada RM highlights this addition to the standard:

http://www.ada-auth.org/standards/aarm12_w_tc1/html/AA-3-5.html